include/clap/ext/draft/scratch-memory.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the plugin request "scratch" memory
+// from the host. Scratch memory can be accessed during the
+// `process()` callback, but is not persistent between callbacks.
+//
+// The motivation for this extension is to allow the plugin host
+// to "share" a single scratch buffer across multiple plugin
+// instances.
+//
+// For example, imagine the host needs to process three plugins
+// in sequence, and each plugin requires 10K of scratch memory.
+// If each plugin pre-allocates its own scratch memory, then 30K
+// of memory is being allocated in total. However, if each plugin
+// requests 10K of scratch memory from the host, then the host can
+// allocate a single 10K scratch buffer, and make it available to all
+// three plugins.
+//
+// On memory-constrained platforms, this optimization may allow for
+// more plugins to be used simultaneously. On platforms with lots
+// of memory, this optimization may improve CPU cache usage.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_SCRATCH_MEMORY[] = "clap.scratch-memory/1";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct clap_host_scratch_memory {
+   // Asks the host for certain amount of scratch memory.
+   // If the host is unable to provide the memory, it should
+   // return "false".
+   //
+   // Note that any memory the host allocates to satisfy
+   // the requested scratch size can be de-allocated by the
+   // host when the plugin is de-activated.
+   //
+   // In the context of plugins and hosts that implement
+   // the "thread-pool" extension, scratch memory is assumed
+   // to be "thread-local". The plugin should request the maximum
+   // amount of scratch memory that it will need on a single
+   // thread. Accordingly, the host must ensure that each
+   // thread can independently provide the requested amount
+   // of scratch memory.
+   //
+   // [main-thread & being-activated]
+   bool(CLAP_ABI *reserve)(const clap_host_t *host, uint32_t scratch_size_bytes);
```

Comment:
Maybe one more question: may the plugin ask multiple times (with decreasing scratch_size_bytes) as long as the host returns false?

Reply:
I think that should be allowed... maybe if the host rejects the amount of scratch memory that the plugin initially requests, the plugin can opt for a more memory-efficient algorithm, or something like that.

Reply:
@Trinitou maybe it is too strict?

I'm fine with the plugin asking as many times as it wants.
The last call "wins", meaning that if the last call failed, no memory is reserved, if the last call succeed then the reserved memory is the one from the last call.

Though, let's add a comment saying that this function is expensive so the plugin should avoid unnecessary calls.

Reply:
> Though, let's add a comment saying that this function is expensive so the plugin should avoid unnecessary calls.

I agree, you're right. That can work as well, totally. 👍 

I just thought that maybe it's more clear (and misbehaviour-checkable) to have more strict constrains. E.g. a plugin validator would have to set an arbitrary limit of `reserve`-calls to detect plugins asking too often.

Maybe both? Strict limitation + comment saying that this function is expensive so the plugin should avoid unnecessary calls

But as I said, I'm fine with both

Reply:
I think I'm against the "strict limitation", but this does bring up another point: The documentation needs to clarify that if the plugin calls `reserve()` multiple times, then the amount of available scratch memory will be the amount specified in the final `reserve()` call (assuming it returned true). I'll have to think about the best way to word that...

Reply:
“The amount of memory available will have size of the last successful call to reserve”

Reply:
If the last call fails then it is 0.

Reply:
Do we need a call to release the reserved memory?

Reply:
> Do we need a call to release the reserved memory?
no - it is valid until deactivation. On activation, the plugin needs to reserve again.